pennylane/beta/tapes/qnode.py

```diff
@@ -0,0 +1,541 @@
+              through the QNode. The QNode accepts and returns Torch tensors.
+
+            * ``interface='tf'``: Allows TensorFlow in eager mode to backpropogate
+              through the QNode. The QNode accepts and returns
+              TensorFlow ``tf.Variable`` and ``tf.tensor`` objects.
+
+            * ``None``: The QNode accepts default Python types
+              (floats, ints, lists) as well as NumPy array arguments,
+              and returns NumPy arrays. It does not connect to any
+              machine learning library automatically for backpropagation.
+
+        diff_method (str, None): the method of differentiation to use in the created QNode.
+
+            * ``"best"``: Best available method. Uses classical backpropagation or the
+              device directly to compute the gradient if supported, otherwise will use
+              the analytic parameter-shift rule where possible with finite-difference as a fallback.
+
+            * ``"backprop"``: Use classical backpropagation. Only allowed on simulator
+              devices that are classically end-to-end differentiable, for example
+              :class:`default.tensor.tf <~.DefaultTensorTF>`. Note that the returned
+              QNode can only be used with the machine-learning framework supported
+              by the device; a separate ``interface`` argument should not be passed.
+
+            * ``"reversible"``: Uses a reversible method for computing the gradient.
+              This method is similar to ``"backprop"``, but trades off increased
+              runtime with significantly lower memory usage. Compared to the
+              parameter-shift rule, the reversible method can be faster or slower,
+              depending on the density and location of parametrized gates in a circuit.
+              Only allowed on (simulator) devices with the "reversible" capability,
+              for example :class:`default.qubit <~.DefaultQubit>`.
+
+            * ``"device"``: Queries the device directly for the gradient.
+              Only allowed on devices that provide their own gradient rules.
+
+            * ``"parameter-shift"``: Use the analytic parameter-shift
+              rule where possible, with finite-difference as a fallback.
+
+            * ``"finite-diff"``: Uses numerical finite-differences for all parameters.=
+
+    Keyword Args:
+        h=1e-7 (float): Step size for the finite difference method.
+        order=1 (int): The order of the finite difference method to use. ``1`` corresponds
+            to forward finite differences, ``2`` to centered finite differences.
+
+    **Example**
+
+    >>> from pennylane.beta.queuing import expval, var, sample, probs
+    >>> from pennylane.beta.tapes import qnode
+    >>> dev = qml.device("default.qubit", wires=1)
+    >>> @qnode(dev)
+    >>> def circuit(x):
+    >>>     qml.RX(x, wires=0)
+    >>>     return expval(qml.PauliZ(0))
+    """
+
+    @lru_cache()

```

Comment:
Why do we use lru_cache here? Does it lead to a big saving?

Reply:
I don't know! It's been in the codebase for as long as I can remember haha

I suppose the thinking is, if you call the decorator again with the same quantum function, it should simply returned the pre-cached QNode, rather than constructing it again? But I'm not sure if this even works.

Reply:
Is the decorator called when when the qfunc is declared, or when it is called? I'm not sure this is a stage of computation that needs to be "sped up"